src/Main.elm

```diff
@@ -1,10 +1,829 @@
+import Task
+import Time
 
 
-main : Html msg
+
+-- we could perhaps model this so that you can only make a "FolderClicked" Msg
+-- or "SortBy" Msg only after the model has loaded up, but I don't think this is
+-- too much of an issue
+
+
+type Msg
+    = FolderClicked Int
+    | GoBack
+    | SortBy Data.Audience.AudienceType
+    | GotAudiencesJSON String
+    | GotAudienceFoldersJSON String
+    | GotTime Time.Posix
+
+
+
+-- we use dicts instead of lists because it gives us O(log n) time complexity
+-- for insertion, removal and querying, as opposed to lists which would be O(n).
+-- additional note: I chose to store audiences and audienceFolders separately because:
+--    1. I don't know if on the server side the IDs of the Folders
+--       and the IDs of the Audiences are kept separately. This way I
+--       would make sure that there's no collisions between them
+--    2. These two come from different API endpoints.
+--       Audiences come from "/api/audiences" and audienceFolders come
+--       from "api/audience_folders".
+--       If we kept these as the same Dict, and only one result would come back,
+--       the dict would look completely normal, but it wouldn't be. Things would
+--       be missing.
+--       So this way, it's either "everything came back, and we render the ui,
+--       or we render some loading page or some error page "
+
+
+type Model
+    = Loading LoadingModel
+    | Loaded LoadedModel
+    | JsonError Json.Decode.Error
+
+
+type alias LoadingModel =
+    { audiences : Maybe (Dict.Dict Int FormattedAudience)
+    , audienceFolders : Maybe (Dict.Dict Int FormattedAudienceFolder)
+    , time : Maybe Time.Posix
+    }
+
+
+type alias LoadedModel =
+    { audiences : Dict.Dict Int FormattedAudience
+    , audienceFolders : Dict.Dict Int FormattedAudienceFolder
+
+    -- we store the full path, because we want to record the full history
+    -- to go back in the tree of folders
+    , currentFolderPath : List Int
+
+    -- according to "SECOND_STEP.md" there must be at least one category selected
+    -- at any time
+    , sortBy : Data.Audience.AudienceType
+    }
+
+
+
+-- we fake an http request with a delay
+
+
+fakeHttpGet : Msg -> Float -> Cmd Msg
+fakeHttpGet msg time =
+    Task.perform (\_ -> msg) (Process.sleep time)
+
+
+init : () -> ( Model, Cmd Msg )
+init flags =
+    let
+        model_ =
+            Loading
+                { audiences = Nothing
+                , audienceFolders = Nothing
+                , time = Nothing
+                }
+
+        -- "load" these at different times, to show that the UI works right
+        -- the user can't do anything with the audienceFolders and audiences
+        -- until both of them load
+        cmd_ =
+            Cmd.batch
+                [ fakeHttpGet (GotAudiencesJSON Data.Audience.audiencesJSON) 300
+                , fakeHttpGet (GotAudienceFoldersJSON Data.AudienceFolder.audienceFoldersJSON) 2000
+                ]
+    in
+    ( model_, cmd_ )
```

Comment:
There is no need for the underscored suffix in here, but I like it as it can be easily forgotten that you are not working with the "real" models and cmds. Or has there been another thinking behind the `_` suffix?